examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -200,5 +226,36 @@ bool NativeStackTrace::error_occured() const {
   return error_occurred;
 }
 
+void NativeStackTrace::prune_dead_pid(uint32_t dead_pid) {
+  auto it = procSymbolsCache.find(dead_pid);
+  if (it != procSymbolsCache.end()) {
+    procSymbolsCache.erase(it);
+  }
+}
+
+void NativeStackTrace::enable_dso_reporting() {
+  NativeStackTrace::insert_dso_name = true;
+}
+
+ProcSyms* NativeStackTrace::get_proc_symbols(uint32_t pid) {
+  static struct bcc_symbol_option symbol_options = {
+    .use_debug_file = 1,
+    .check_debug_file_crc = 1,
+    .lazy_symbolize = 1,
+    .use_symbol_type = BCC_SYM_ALL_TYPES
+  };
+  auto current_time_point = std::chrono::steady_clock::now();
+  double timestamp_s = std::chrono::duration_cast<std::chrono::duration<double>>(current_time_point - startup_time_point).count();
+  if (!procSymbolsCache.count(pid)) {
+    procSymbolsCache[pid] = ProcSymbolsCacheEntry{timestamp_s, unique_ptr<ProcSyms>(new ProcSyms(pid, &symbol_options))};
+  } else {
+    if (timestamp_s - procSymbolsCache[pid].timestamp_s > ProcSymbolsCacheTTL_S) {
+      procSymbolsCache[pid].proc_syms->refresh();
```

Comment:
I looked at its implementation, very inefficient :/ It flushes all loaded modules and re-loads them, instead of removing modules that have changed loading new ones.

Well, we'll talk about performance after CR

Reply:
Btw - will it handle if the PID changes? i.e let's say we had python with PID=55. Then 10 minutes passed and now PID=55 is a different Python application. Will `refresh` work fine?

Reply:
(or, do we expect `prune_dead_pid` to prune the dead PIDs before they are being recycled)

Reply:
That's a valid concern.
We might run into 1 minute of stale data, with the current refresh implementation.
`prune_dead_pid` might not help with pid reuse issue.
We might need to research this a bit.
I can think of two approaches:
- improve identification of processes - use more information, perhaps process start_time?, and store it in `PidData`,
- register a process exit hook with eBPF to handle process departure.
The first one might be simpler to take.

Reply:
The first approach is simpler to take, but anyway - `updateProcesses`'s default interval is 10 seconds, so `populatePidTable` is called every 10s and thus `prune_dead_pid`. I think 10s is good enough - I don't expect PID reuse in such a short timeframe, it's extreme

Reply:
Resolving this